File: 907-koko-eating-bananas/koko-eating-bananas.cpp
class Solution {
public:
    bool is_possible(vector<int>& piles, long long int h, long long int mid){
        int n =piles.size();
        long long int sum = 0;
        
        for(int i=0; i < n; i++)
        {
            long long int cnt = ceil ((double)piles[i]/ (double)mid);
            sum += cnt;
        }
        cout<<sum<<" ";
        if(sum <= h){
            return true;
        }
        return false;
    }

    int minEatingSpeed(vector<int>& piles, long long int h) {
        int n = piles.size();
        int maxi = *max_element(piles.begin(),piles.end());
        
        int low = 1; 
        int high = maxi;
        int ans = INT_MAX;

        while(low <= high){
           long long int mid = (low + high )/2;

            if(is_possible(piles, h, mid)){
                ans = mid;
                // cout<<ans<<" ";
                high = mid - 1;
            }
            else{
                low = mid + 1;
            }
        }
        return ans;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public boolean canEat(int[] piles, int h, int k){
        int n = piles.length;
        int totalHrs = 0;

        for(int i=n-1; i>=0; i--){
            double timeRequire = Math.ceil((double)piles[i]/k);
            totalHrs += timeRequire;
            if(totalHrs > h) return false;
        }

        return true;
    }

    public int minEatingSpeed(int[] piles, int h) {
        int n = piles.length;
        Arrays.sort(piles);

        if(h < n) return -1;

        int left = 1;
        int right = piles[n-1];

        while(left < right){
            int mid = left + (right - left)/2;
            if(canEat(piles, h, mid)){
                right = mid;
            }else{
                left = mid + 1;
            }
        }

        return right;
    }
}